src/include/simeng/arch/riscv/Instruction.hh

```diff
@@ -147,36 +158,63 @@ class Instruction : public simeng::Instruction {
   /** Is this a branch operation? */
   bool isBranch() const override;
 
-  /** Is this an atomic instruction? */
-  bool isAtomic() const;
-
-  /** Is this a floating point operation? */
-  bool isFloat() const;
-
   /** Retrieve the instruction group this instruction belongs to. */
   uint16_t getGroup() const override;
 
-  /** Set this instruction's execution information including it's execution
-   * latency and throughput, and the set of ports which support it. */
-  void setExecutionInfo(const executionInfo& info);
+  /** Check whether all operand values have been supplied, and the instruction
+   * is ready to execute. */
+  bool canExecute() const override;
+
+  /** Execute the instruction. */
+  void execute() override;
 
   /** Get this instruction's supported set of ports. */
   const std::vector<uint16_t>& getSupportedPorts() override;
 
+  /** Set this instruction's execution information including it's execution
+   * latency and throughput, and the set of ports which support it. */
+  void setExecutionInfo(const ExecutionInfo& info) override;
+
   /** Retrieve the instruction's metadata. */
   const InstructionMetadata& getMetadata() const;
 
   /** Retrieve the instruction's associated architecture. */
   const Architecture& getArchitecture() const;
 
-  /** The maximum number of source registers any supported RISC-V instruction
-   * can have. */
-  static const uint8_t MAX_SOURCE_REGISTERS = 3;
-  /** The maximum number of destination registers any supported RISC-V
-   * instruction can have. */
-  static const uint8_t MAX_DESTINATION_REGISTERS = 1;
+  /** Retrieve the identifier for the first exception that occurred during
+   * processing this instruction. */
+  InstructionException getException() const;
 
  private:
+  /** Process the instruction's metadata to determine source/destination
+   * registers. */
+  void decode();
+
+  /** Update the instruction's identifier with an additional field. */
+  void setInstructionIdentifier(InsnIdentifier identifier) {
+    instructionIdentifier_ |= static_cast<uint16_t>(identifier);
+  }
+
+  /** Test whether this instruction had the given identifier set. */
+  constexpr bool isInstruction(InsnIdentifier identifier) const {
+    return (instructionIdentifier_ & static_cast<uint16_t>(identifier));
+  }
+
+  /** For instructions with a valid rm field, extract the rm value and change
+   * the CPP rounding mode accordingly, then call the function "operation"
+   * before reverting the CPP rounding mode to its initial value. "Operation"
+   * should contain the entire execution logic of the instruction */
+  void setStaticRoundingModeThen(std::function<void(void)> operation);
+
+  /** Generate an ExecutionNotYetImplemented exception. */
+  void executionNYI();
+
+  /** Is this an atomic instruction? */
+  bool isAtomic() const;
```

Comment:
Why have `isAtomic` and `isFloat` moved to private?

Reply:
They are only used internally in the instruction class, but they can probably be removed completely